analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/SymbolicValue.cs

```diff
@@ -72,5 +90,111 @@ public sealed record SymbolicValue
             Constraints.Any()
                 ? Constraints.Values.Select(x => x.ToString()).OrderBy(x => x).JoinStr(", ")
                 : "No constraints";
+
+        private SymbolicValue RemoveConstraint(Type type) =>
+            Constraints.Count switch
+            {
+                1 => Empty,
+                2 => OtherSingle(type),
+                3 => OtherPair(type),
+                _ => this with { Constraints = Constraints.Remove(type) },
+            };
+
+        private SymbolicValue OtherSingle(Type except)
+            => SingleConstraintValue(OtherSingleConstraint(except));
+
+        private SymbolicConstraint OtherSingleConstraint(Type except)
+        {
+            SymbolicConstraint otherConstraint = null;
+            foreach (var kvp in Constraints)
+            {
+                if (kvp.Key != except)
+                {
+                    otherConstraint = kvp.Value;
+                    break;
+                }
+            }
+
+            return otherConstraint;
+        }
+
+        private SymbolicValue OtherPair(Type except)
+        {
+            SymbolicConstraint first = null;
+            SymbolicConstraint second = null;
+            foreach (var kvp in Constraints)
+            {
+                if (kvp.Key != except)
+                {
+                    if (first == null)
+                    {
+                        first = kvp.Value;
+                    }
+                    else
+                    {
+                        second = kvp.Value;
+                        break;
+                    }
+                }
+            }
+
+            return PairConstraintValue(first, second);
+        }
+
+        private static SymbolicValue SingleConstraintValue(SymbolicConstraint constraint)
+        {
+            var cacheKey = new CacheKey(constraint.Kind);
+            if (cache.TryGetValue(cacheKey, out var result))
+            {
+                return result;
+            }
+            else
+            {
+                result = Empty with { Constraints = Empty.Constraints.SetItem(constraint.GetType(), constraint) };
+                return cache.GetOrAdd(cacheKey, result);
+            }
+        }
+
+        private static SymbolicValue PairConstraintValue(SymbolicConstraint first, SymbolicConstraint second)
```

Comment:
For the cache entry for the pair, I need to fetch the single entry and then construct the new one (see how `single` is accessed twice for the `with`):
```cs
                var single = CachedSymbolicValue(first);
                result = single with { Constraints = single.Constraints.SetItem(second.GetType(), second) };
```
If we do it in one method, I probably need some kind of recursive call for this part. It's possible to do, but I don't really see the gain. I also don't want to do Empty -> First -> First+Second because the "First" would not come from the cache in that case.

Reply:
I gave it a try and it looks like this, instead of 2 methods:
```
        private static SymbolicValue CachedSymbolicValue(SymbolicConstraint first, SymbolicConstraint second = null)
        {
            // Performance: Don't use the factory overload of GetOrAdd
            var cacheKey = new CacheKey(first.Kind, second?.Kind);
            if (cache.TryGetValue(cacheKey, out var result))
            {
                return result;
            }
            else
            {
                if (second is null)
                {
                    result = Empty with { Constraints = Empty.Constraints.SetItem(first.GetType(), first) };
                }
                else
                {
                    result = CachedSymbolicValue(first);
                    result = result with { Constraints = result.Constraints.SetItem(second.GetType(), second) };
                }
                return cache.GetOrAdd(cacheKey, result);
            }
        }
```
When constructing the pair, there's no big difference if `CachedSymbolicValue(first)` calls itself, or the other almost duplicate overload.
Like this, the main caching logic is there only once. And the implementation detail of building `single` vs `pair` is handled by the `if`